### ISSUE TITLE
Making an Alarm in MAS

### ISSUE THREAD

**Issue by DeputyLOL** [MOARdV/AvionicsSystems#18]
Sorry about the constant silly questions, once I get the hang of it... there will probably still be loads of questions, just not as dumb.
I need to make an alarm for my IVA. It doesn't need a light or anything, just a sound to play.

... how do I do that?

**Comment by MOARdV** [comment]
On a prop - preferably one that is installed only one time in the IVA, add a MASComponent with an AUDIO_PLAYER action.  (If you already have a prop that's used once, you can just add the AUDIO_PLAYER to it)  The "documentation" at [here](https://github.com/MOARdV/AvionicsSystems/blob/master/Documents/MASComponent.txt) is hopefully adequate for configuring it.

**Comment by DeputyLOL** [comment]
thank you

**Comment by DeputyLOL** [comment]
I have a weird bug. I have too separate monitors with separate pages, but both pages are being displayed on both monitors.

**Comment by MOARdV** [comment]
> I have a weird bug. I have too separate monitors with separate pages, but both pages are being displayed on both monitors.

Okay,  I came across the problem a couple of weeks ago, and I only partially fixed it, since all I had was text at the time.  I'll get that fixed today, hopefully.

**Comment by DeputyLOL** [comment]
When I try and use fc.AngleOfAttack() I get this:
[ERR 16:38:40.338] [AvionicsSystems] INITIALIZATION ERROR: Error creating variable ACK320SpeedAlphaProt()

[ERR 16:38:40.338] [Variable] Unknown variable 'ACK320SpeedAlphaProt()':

[ERR 16:38:40.339] [Variable] MoonSharp.Interpreter.ScriptRuntimeException: cannot access field AngleOfAttack of userdata<AvionicsSystems.MASFlightComputerProxy>
  at MoonSharp.Interpreter.Execution.VM.Processor.ExecIndex (MoonSharp.Interpreter.Execution.VM.Instruction i, Int32 instructionPtr) [0x00000] in <filename unknown>:0 
  at MoonSharp.Interpreter.Execution.VM.Processor.Processing_Loop (Int32 instructionPtr) [0x00000] in <filename unknown>:0 

[LOG 16:38:40.341] [MASFlightComputer] Adding new variable 'ACK320SpeedAlphaProt()'

Is this because angle of attack isn't in the dev build yet?

**Comment by MOARdV** [comment]
It should be in the dev build I posted yesterday (the one linked to on this wiki)

**Comment by DeputyLOL** [comment]
That's the one I'm using....

**Comment by DeputyLOL** [comment]
I tried using fc.AngleOfAttack() directly and I got this:
[EXC 21:57:16.947] ArgumentException: method return type is incompatible
	System.Delegate.CreateDelegate (System.Type type, System.Object firstArgument, System.Reflection.MethodInfo method, Boolean throwOnBindFailure)
	System.Delegate.CreateDelegate (System.Type type, System.Reflection.MethodInfo method, Boolean throwOnBindFailure)
	System.Delegate.CreateDelegate (System.Type type, System.Reflection.MethodInfo method)
	AvionicsSystems.MASIFAR..cctor ()
	Rethrow as TypeInitializationException: An exception was thrown by the type initializer for AvionicsSystems.MASIFAR
	AvionicsSystems.MASFlightComputer.Start ()
[ERR 21:57:16.951] [MASFlightComputer] MASFlightComputer.FixedUpdate exception: System.NullReferenceException: Object reference not set to an instance of an object
  at AvionicsSystems.MASFlightComputer.FixedUpdate () [0x00000] in <filename unknown>:0

**Comment by MOARdV** [comment]
Stupid copy/paste error I overlooked.  It's fixed now.

**Comment by DeputyLOL** [comment]
Thanks!
I still have the '2 separate pages on both monitors' issue though.

**Comment by MOARdV** [comment]
Yeah, I haven't gotten to the separate pages issue.

**Comment by MOARdV** [comment]
I think the separate pages issue is fixed - I don't have multiple monitors available that use non-text page items in my development environment to verify.

**Comment by DeputyLOL** [comment]
How do i do "less than or equal to"? is it <= or =<?
Or do i have to do BLAH < BLAHBLAH or BLAH = BLAHBLAH?

**Comment by DeputyLOL** [comment]
For the audio player, is there some way to make the whole sound clip play?
I'm using it for altitude callouts but the sound gets cut off if you decsend too fast.

**Comment by MOARdV** [comment]
> How do i do "less than or equal to"? is it <= or =<?

`<=`

> For the audio player, is there some way to make the whole sound clip play?

Not yet.

**Comment by DeputyLOL** [comment]
Should i start a new issue for the audio features/suggestions

**Comment by MOARdV** [comment]
> Should i start a new issue for the audio features/suggestions

Issue #9 is already tracking some audio features - you can add to it.

**Comment by DeputyLOL** [comment]
Quick question...
How do i create a JSINumericalInput switch in MAS?

**Comment by MOARdV** [comment]
> How do i create a JSINumericalInput switch in MAS?

`fc.AddPersistent("persistentName", amount)`.

Take a look at the props in AvionicsSystems/ASET_Props/swBKnob (the brightness control knobs).

**Comment by DeputyLOL** [comment]
Ahhh.... Thank you!

**Comment by DeputyLOL** [comment]
is there some way to set a default value?

**Comment by MOARdV** [comment]
> is there some way to set a default value?

Yes.  In your part definition where you have the MASFlightComputer module, add:

```
PERSISTENT_VARIABLES
{
	...
}
```

Where the `...` are the named persistents and initializers (eg, `Backlight = 0`), one per line.

**Comment by DeputyLOL** [comment]
Thank you. This will be useful. 
Another quick question. Is it possible to do state variables and time delays?

**Comment by MOARdV** [comment]
> Is it possible to do state variables and time delays?

State variables would need to be done in your Lua scripting.  Take a look in AvionicsSystems/ASET_Props/ARRT/ARRT.lua for one state variable example, and in DSKY/DSKY.lua for a much more complex example (that I think is still incomplete).  Or, for that matter, ClockTimer/ASET_ClockTimer.lua.  The latter has a couple of things I need to update, since it was one of the earliest props I converted, but it is still fully functional.

Time delays can be done as well, since the Lua script can access MET or UT, so it can determine elapsed time (the clock timer scripting shows a way to do that).

**Comment by DeputyLOL** [comment]
Thanks, ill take a look now!

**Comment by DeputyLOL** [comment]
I apologize for this but it all goes over my head. Once i understand it better I will be able to use it properly. Do you think you could explain how it works?

**Comment by MOARdV** [comment]
> Do you think you could explain how it works?

It's been a busy month ... :(

What part(s) should I explain?  I think the ASET_ClockTimer.lua will be the easiest to work with - is has a state machine (for the stopwatch timer) as well as something like a time delay.   Also give me an idea of what you're trying to accomplish, so I can focus on what you're trying to do, instead of explaining stuff that doesn't help.

**Comment by DeputyLOL** [comment]
Sorry for inactivity. University turns out to be very time consuming. 
Who'd a' thunk it?
I need a state machine for the master caution and master warning system.
If theres a problem, turn the warning light on. 
If i cancel it, turn the warning light off.
If the problem disappears on its own, turn the warning light off.
If another problem occurs after the warning light has been cancelled, turn the light on again.
The light needs to cancelled again after it turns on again.

**Comment by MOARdV** [comment]
> Sorry for inactivity. University turns out to be very time consuming. 
> Who'd a' thunk it?

Funny how that works, eh?

I'll work up some sample scripts that should work for that scenario.

**Comment by DeputyLOL** [comment]
thanks. 
just from brain storming it, i came to the conclusion that every different alarm would need its own state.
does that sound right?

**Comment by MOARdV** [comment]
> just from brain storming it, i came to the conclusion that every different alarm would need its own state.
does that sound right?

Maybe.  It depends on whether you're willing to accept some degree of "cheating" in the alarm state tracking, and how many overlapping simultaneous alarms you might have at a given time.

One way to handle the alarm snooze function would be to assign each alarm a different number.  Say "No alarm" = 0, "Too low" = 1, "Stall" = 2, "Overspeed" = 3, "Lower Landing Gear!" = 4, etc, ad nauseum.  You could prioritize the alarms, so larger numbers are more important.

When an alarm is active, it sets a `currentAlarm` variable to the larger of `currentAlarm` or its current value.  When the 'snooze' button is hit, it sets `ignoreAlarm` to the `currentAlarm` value.  To decide if any alarm is active, you check if `currentAlarm` is > 0.  To decide if you need the audible signal, you test for `ignoreAlarm` < `currentAlarm`.  That may not be clear - I'm thinking 'out loud'.  I'll try to write it out as a script to see I can make it clearer.  There are probably some situations that need to be thought out, anyway, that I haven't thought about.

**Comment by DeputyLOL** [comment]
i think it makes sense, but most of my alarms have the same 'priority'.
I did more brain storming and my brain cobbled together this:
There is a 'master alarm (master)' and a bunch of 'sub alarms (subs)'.
The master will turn on if any of the subs are true. At first it would make sense to put the state variable in master, but what if you have a cancelled sub and another sub becomes true?!?!?!?!
Instead of having the state in the master you could have them in the states.

AC_randomsubalarmthingy
{
if (criteria == 1)
 {
 return 1;
 while(criteria == 1 && snooze == 1)
 {
  return 0;
 }
}
else
{
 return 0;
}
}

WHOOO IM LEARNEN SHIT!!!!!!

The code above would be used in a sub. 
It would return 1 if the criteria is met.
If the snooze button if pressed while the criteria is met then it will return 0 until the criteria is not met, then met again.

**Comment by MOARdV** [comment]
I think what you want would be closer to

```
AC_randomsubalarmthingy()
{
  if (criteria == 1) // problem is active
  {
    if (snooze == true) // Snooze button was pressed
    {
      return 0; // Don't nag the pilot
    }
    else
    {
      masterAlarm = true; // See below...
      return 1; // HEY!  BAD THINGS ARE HAPPENING!
    }
  }
  else // No alarm
  {
    snooze = false; // Reset the snooze
    return 0;
  }
}
```

This will automatically reset the snooze for that particular alarm when the alarm shuts off (it's no big deal to repeatedly shut it off).  The down side is you will need to have a snooze variable for each alarm type, and the snooze button action will look like this:

```
AC_AlarmACK
{
  snoozeOverspeed = true;
  snoozeTooLow = true;
  snoozeTooFast = true;
  ... etc
}
```

It's actually not a big deal for the button to set lots of variables, since the performance cost of that action only applies when the button is actually pressed.

The tricky part is deciding if the master alarm is on.  Well, actually "alarm on" is easy - anywhere the sub alarm does a return 1, you also set `masterAlarm = true`, like I did above.  The trick is knowing when you can shut it off.  I think this might actually work:

```
AC_masterAlarmOn
{
  if (masterAlarm == true)
  {
    masterAlarm = false; // Reset it for next update.
    return 1;
  }
  else
  {
    return 0;
  }
}
```

This way, every update, if the alarm is on, `AC_masterAlarmOn` will say "yep" if the `masterAlarm` variable was set.  It will then clear the master alarm flag, and next FixedUpdate, when the sub alarms decide if they need to yell, the sub alarms will be able to set the master alarm flag.

**Comment by DeputyLOL** [comment]
ok.. i think i get it. 
The reason why i had the while loop is because the snooze button isn't an off/on switch, its a momentary button. My thinking was that it would trap the alarm in a loop if the snooze button was ever on and the only way the alarm could turn on again was if the criteria goes away(thus breaking the loop) and then comes back.

**Comment by MOARdV** [comment]
> The reason why i had the while loop is because the snooze button isn't an off/on switch, its a momentary button.

The while loop would never actually execute, since the `return` right before it would end that code block.  Besides which, a while loop would effectively hang the script (other than it also have a return), so control would never pass back to MAS and KSP.  While loops are Bad Things in event-driven programming (which is what this is).

If the idea is that the snooze button is momentary, then you can remove the `snooze = false` from that function, and instead use this for `onClick` for the snooze button:
```
AC_SnoozeAlarmClick()
{
  snooze = true;
}
```

and this for `onRelease` for the same button
```
AC_SnoozeAlarmRelease()
{
  snooze = false;
}
```

**Comment by DeputyLOL** [comment]
Ok, this looks good. Will it work for multiple alarms though? 
If an alarm goes off and I presse the snooze button it should only inhibit that alarm.

**Comment by MOARdV** [comment]
> If an alarm goes off and I presse the snooze button it should only inhibit that alarm.

As presented, it assumes a single snooze button that silences all alarms as long as the button is held.  If it should only snooze the alarms that are active at the time it was pressed, but not any new ones that trigger while it is pressed, you would need to have a convoluted series of if / elseif to figure out which alarms are active, and independent snooze variables for each of them.  Honestly, that may be accurate, but I don't see it adding much value over a single blanket snooze button.

Now, if the idea is that you press the snooze button, and things stay snoozed when it's released, that would be a different matter, and that goes back to the design I had proposed earlier.

**Comment by DeputyLOL** [comment]
I think im gonna have to push for realism here.
Im still working on it and i think i may have something

**Comment by DeputyLOL** [comment]
a little off topic
how do i do transforms for a MAS switch?

**Comment by MOARdV** [comment]
> how do i do transforms for a MAS switch?

In what sense?  Making a switch in a 3D modeling program, or configuring a collider in MAS, or what?

**Comment by DeputyLOL** [comment]
Sorry, thinking in RPM sense.
How do I get a prop to move X, Y and Z when a variable changes?

**Comment by DeputyLOL** [comment]
Ok, i came up with this:
function ACK320WarningEngineFire()
	local WarningInhibit
	local WarningOn
	if fc.GetPersistent("AC_A320_TEST") == 1 and fc.GetPersistent("AC_A320_WarningCancel") == 1 then  -- CONDITION TRUE AND SNOOZE BUTTON PRESSED
		WarningInhibit = 1
		WarningOn = 0
	elseif fc.GetPersistent("AC_A320_TEST") == 1 and WarningInhibit == 0 then
		WarningOn = 1
	elseif fc.GetPersistent("AC_A320_TEST") == 0 then
		WarningOn = 0
		WarningInhibit = 0
	else
		WarningOn = 1
	end
	if WarningOn == 1 then
		return 1
	else
		return 0
	end
end

But it doesnt work. Are the local variables reset on every run?

**Comment by MOARdV** [comment]
> How do I get a prop to move X, Y and Z when a variable changes?

`TRANSLATION`, which I apparently forgot to actually document in the MASComponent.txt.  It's basically identical to `ROTATION`, but it uses `startTranslation` and `endTranslation` instead of `startRotation` and `endRotation`.

> But it doesnt work. Are the local variables reset on every run?

Because they are inside the function, yes.  Move the two locals outside of the function, and give them initial values (like 0), and they will persist between calls:

``` Lua
local WarningInhibit = 0
local WarningOn = 0
function ACK320WarningEngineFire()
  ...
end
```


Once you have it working to your satisfaction, show me the code, and I'll show you some optimization tricks to make it a little less costly to execute.

**Comment by DeputyLOL** [comment]
thank you

**Comment by DeputyLOL** [comment]
its still not working... 
It would be very helpful to see the local variables but i suppose thats why they are called local.

**Comment by DeputyLOL** [comment]
my issue is that the 'inhibit' is not turning to 1

**Comment by DeputyLOL** [comment]
Im really confused. If the first if statement is true, shouldn't it skip all the else ifs?

**Comment by MOARdV** [comment]
A couple of other things while I look at that function again:

* First, always use `GetPersistentAsNumber()`, not `GetPersistent`, if you expect a number - if a persistent value doesn't exist yet, the first one will return 0, the second will return an invalid value.  That may also explain some problems you had with other scripts.  `GetPersistent` is needed because a persistent may be a number or a string, and in certain cases you don't care which it is.

* Second: you can use `fc.LogMessage("Some text")` to write your own messages into the log file.  The Lua string library will give you access to formatting and stuff, like `fc.LogMessage(string.format("WarningInhibit = %d, WarningOn = %d", WarningInhibit, WarningOn))`.

Add some fc.LogMessage lines to see which code path is executing, and change `GetPersistent` to `GetPersistentAsNumber` everywhere you use it (unless you know it is supposed to be a string value).

**Comment by DeputyLOL** [comment]
Ok, will try that thanks

**Comment by DeputyLOL** [comment]
Ok, still not working...
The problem is that the first if statement is apparently never true. But, there are only 2 conditions and BOTH ARE WORKING!!!!! 
I know this is getting repetitive but do you understand what its suppose to do and can you tell me why it isn't doing it?

**Comment by MOARdV** [comment]
> I know this is getting repetitive but do you understand what its suppose to do and can you tell me why it isn't doing it?

Let me tell you what I think you're trying to do, to see if I understand it correctly.

You have 3 variables that control the alarm: `AC_A320_TEST` - this is the "alarm condition" variable that indicates "BAD THINGS".  `AC_A320_WarningCancel` - this is what I was calling the "snooze" variable that the pilot uses to acknowledge the alarm so it will shut up.  `WarningInhibit` - this is the local variable that is used to remember that the pilot already acknowledged the error, so he doesn't have to hold down the snooze button.

What is supposed to happen is this:

* When an alarm is signaled, the warning light / sound triggers, and the pilot is notified.
* When the pilot hits the "snooze" button, the alarm is muted, and it remains muted by the inhibit variable.
* When the alarm is no longer applicable, the alarm is no longer signaled, and the inhibit variable is reset.

So, basically, there are three states: No Alarm, Active Alarm, and Muted Alarm.

Let me rearrange your sample code and add some debug stuff and comments, and see what I come up with.

**Comment by DeputyLOL** [comment]
you got it spot on. 
and i just got it to work... sort of. I put in a second test switch to use as the snooze button. At first it didnt work, then KSP crashed, i reload it and now it does work!?!?!
So maybe its the switch thats the problem. So now the question becomes. what wrong with the switch?

**Comment by MOARdV** [comment]
Okay.  This function breaks the behavior down by considering only one variable at a time, and only if it matters to the current state of the alarm system.  The most important question is "Is there a warning?", so we check it first.  If it's off, then we don't care about anything else - we reset the inhibit value and we're done.  If it's on, then the next question is "did the pilot acknowledge the warning?".  If he did, then we're done once again.  If he did not, then the last question is "Is the pilot now acknowledging the warning?".  If so, we remember that for next time around; if not, we sound the alarm.

```Lua
-- A local is less expensive than a persistent, but it means the snooze button is reset
-- each time the IVA is loaded.  Which only matters if the plane had alarms going off while
-- on the ground or in orbit, and the player switched away from the craft and returned.
-- Since there are multiple alarms, each alarm should have a unique WarningInhibit value,
-- eg, WarningEngineFireInhibit, WarningPullUpInhibit, etc.
local WarningInhibit = false

function ACK320WarningEngineFire()

	-- This can be a local inside the function - we don't care about it outside
	-- of this function, right?
	local WarningOn = 0
	
	if fc.GetPersistentAsNumber("AC_A320_TEST") > 0 then
	
		-- The warning condition is met.  Decide if we tell the pilot:
		if WarningInhibit == false then
		
			if fc.GetPersistentAsNumber("AC_A320_WarningCancel") > 0 then
				-- The pilot pressed the cancel button, but we haven't inhibited the warning yet
				fc.LogMessage("AC_A320_TEST is on, WarningInhibit is false, AC_A320_WarningCancel is on -> set WarningInhibit, alarm off")
				WarningInhibit = true
			else
				-- The pilot has not canceled the warning.  Sound the alarm.
				fc.LogMessage("AC_A320_TEST is on, WarningInhibit is false, AC_A320_WarningCancel is off -> ALARM ON")
				WarningOn = 1
			end
		
		else
			-- This is only for debugging - the 'else' statement and the LogMessage should be removed for production code
			fc.LogMessage("AC_A320_TEST is on, WarningInhibit is true -> alarm off")
		end
		
	else
		-- The warning condition is not met, so reset the inhibit value.
		WarningInhibit = false
		
		fc.LogMessage("AC_A320_TEST is off -> reset WarningInhibit, alarm off")
	end
	
	return WarningOn
end
```

**Comment by MOARdV** [comment]
> So now the question becomes. what wrong with the switch?

The snooze switch?  I would assume it does something like this in the COLLIDER_EVENT:

```
  onClick = fc.SetPersistent("AC_A320_WarningCancel", 1)
  onRelease = fc.SetPersistent("AC_A320_WarningCancel", 0)
```

That sets the persistent on when it's pressed, and off when it's released.

**Comment by DeputyLOL** [comment]
At first it they were both TogglePersistent
Then it was AddPersistentClamped to match the rotary switches and to try and figure out why it wasnt working.
Then it was togglePersistent again when i removed the onRelease for more debugging.

**Comment by DeputyLOL** [comment]
OK I FIXED. I just had to change the name of the persistent to something else apparently!
I didn't think i had a persistent of that name but apparently i did.
Is the code you posted above more efficient then mine?

**Comment by MOARdV** [comment]
> Is the code you posted above more efficient then mine?

It should be - most of the time in normal flight conditions, all alarms should be off, and we don't care if the inhibit button is pressed, so we check the alarm condition persistent value first.  Only if it's active do we need to check anything else.  And we don't need to check the "Cancel Alarm" persistent if the pilot has already inhibited this alarm, so we don't have to incur the cost of a second fc.GetPersistentAsNumber call unless a) the alarm is active, and b) the pilot has not already inhibited it.

**Comment by DeputyLOL** [comment]
Right then, ill use yours. Thanks very much.
Another question though. Can a single prop play multiple sounds at the same time?

**Comment by MOARdV** [comment]
> Can a single prop play multiple sounds at the same time?

Sure.  You can use multiple AUDIO_PLAYER objects, each with a separate sound and/or separate conditions for triggering the sound.

**Comment by DeputyLOL** [comment]
ok.. thats what i thought. but that doesn't appear to happen.

**Comment by MOARdV** [comment]
> ok.. thats what i thought. but that doesn't appear to happen.

Oh, I think I know the issue.  I'm adding the audio source to the prop's gameObject, but you can only add one to an object, so second and subsequent sounds are either doing nothing, or they're overwriting the existing audio object.  It should be easy to fix.  I'll add a reminder.

**Comment by DeputyLOL** [comment]
thank you

**Comment by MOARdV** [comment]
I *think* this issue can be closed - the outstanding issue is supporting multiple audio sources on a single prop, which has its own issue.
